src/ogdf/cluster/ClusterGraph.cpp

```diff
@@ -1186,90 +1244,24 @@ bool ClusterGraph::representsCombEmbedding() const {
 	if (!m_adjAvailable) {
 		return false;
 	}
-
 #ifdef OGDF_DEBUG
 	consistencyCheck();
 #endif
 
-	cluster c = nullptr;
-	forall_postOrderClusters(c, (*this)) {
-#ifdef OGDF_HEAVY_DEBUG
-		Logger::slout()
-				<< "__________________________________________________________________" << std::endl
-				<< std::endl
-				<< "Testing cluster " << c << std::endl
-				<< "Check on AdjList of c" << std::endl;
-		for (adjEntry adjDD : c->adjEntries) {
-			Logger::slout() << adjDD << ";  ";
-		}
-		Logger::slout() << std::endl;
-#endif
-
-		if (c != m_rootCluster) {
-			ListConstIterator<adjEntry> it;
-			it = c->firstAdj();
-			adjEntry start = *it;
-
-#ifdef OGDF_HEAVY_DEBUG
-			Logger::slout() << "firstAdj " << start << std::endl;
-#endif
-
-			while (it.valid()) {
-				AdjEntryArray<bool> visitedAdjEntries((*m_pGraph), false);
-
-				ListConstIterator<adjEntry> succ = it.succ();
-				adjEntry adj = *it;
-				adjEntry succAdj;
-
-				if (succ.valid()) {
-					succAdj = *succ;
-				} else {
-					succAdj = start; // reached the last outgoing edge
-				}
-
-#ifdef OGDF_HEAVY_DEBUG
-				Logger::slout() << "Check next " << std::endl;
-				Logger::slout() << "current in adj list of" << adj << std::endl;
-				Logger::slout() << "succ in adj list of c " << succAdj << std::endl;
-				Logger::slout() << "cyclic succ in outer face " << adj->cyclicSucc() << std::endl;
-#endif
-
-
-				if (adj->cyclicSucc() != succAdj)
-				// run along the outer face of the cluster
-				// until you find the next outgoing edge
-				{
-					adjEntry next = adj->cyclicSucc();
-					adjEntry twin = next->twin();
-
-#ifdef OGDF_HEAVY_DEBUG
-					Logger::slout() << "Running along the outer face ... " << std::endl;
-					Logger::slout() << "next adj " << next << std::endl;
-					Logger::slout() << "twin adj " << twin << std::endl;
-#endif
-
-					if (visitedAdjEntries[twin]) {
-						return false;
-					}
-					visitedAdjEntries[twin] = true;
-					while (next != succAdj) {
-						next = twin->cyclicSucc();
-						twin = next->twin();
-#ifdef OGDF_HEAVY_DEBUG
-						Logger::slout() << "Running along the outer face ... " << std::endl;
-						Logger::slout() << "next adj " << next << std::endl;
-						Logger::slout() << "twin adj " << twin << std::endl;
-#endif
-						if (visitedAdjEntries[twin]) {
-							return false;
-						}
-						visitedAdjEntries[twin] = true;
-					}
+	// FIXME this fails for disconnected graphs
+	for (cluster c = firstPostOrderCluster(); c != m_rootCluster; c = c->pSucc()) {
+		for (auto it = c->adjEntries.begin(); it.valid(); it++) {
+			adjEntry adj = *it;
+			adjEntry succAdj = *c->adjEntries.cyclicSucc(it);
+
+			// run along the outer face of the cluster until you find the next outgoing edge
+			adjEntry next = adj->cyclicSucc();
+			while (next != succAdj) {
+				if (next == adj->twin()) {
```

Comment:
If the ClusterGraph is really messed up and does not represent a combinatorial embedding (more specifically: when neither `succAdj` nor `adj->twin()` is ever reached), couldn't this end up in an infinite loop? If you tell me that cannot happen, I'll just believe you, though. :D

Reply:
I guess that you'd still need to return to the initial adjEntry at some point, but I'm also not 100% sure that you couldn't produce some weird situation where this doesn't happen. Added a counter to guard against this in debug mode, as you cannot visit an edge more than twice in a valid embedding.